mpdev_update.1
.TH mpdev_update 1 "December 23, 2009" "manual"
.SH NAME
.PP
mpdev_update - Update stats, sticker sqlite db from mpd database
.SH SYNOPSIS
.PP
mpdev_update [\f[I]option\f[]]

.SH DESCRIPTION
.PP
\fBmpdev_update\fR creates/updates stats.db or sticker.db from the
\fBmpd\fR(1) database. It sends the \fIlistinfo\fR command to mpd to
get list of all music files. \fBmpdev_update\fR uses the \fI%mtime%\fR
tag for a file from the mpd(1) database.

.SH OPTIONS
.TP 3
.B -v
Set verbose output. You can increase verbosity by specifying it multiple times
.RS
.RE

.TP 3
\fB-i\fR \fIIP\fR
IP address of mpd host if different from localhost
.RS
.RE

.TP 3
\fB-p\fR \fIport\fR
port on which mpd(1) is listening if different from 6600
.RS
.RE

.TP 3
\fB-s\fR \fIunix_socket\fR
Unix domain socket in case you don't want to use \fIIP\fR and \fIport\fR
.RS
.RE

.TP 3
\fB-d\fR \fIdb_file_path\fR
full path to the sqlite3(1) stats database
.RS
.RE

.TP 3
\fB-D\fR \fI0\fR | \fI1\fR
0 to insert new records in stats.db, 1 to insert new records in sticker.db
.RS
.RE

.TP 3
.B -U
Use update mode. The default is to use insert mode.
.RS
.RE

.TP 3
\fB-j\fR 
Puts the journal in volatile memory. This saves disk I/O but at the expense
of database safety and integrity. If the application using SQLite crashes
in the middle of a transaction when the MEMORY journaling mode is set, then
the database file will very likely go corrupt.
.RS
.RE

.TP 3
\fB-t\fR 
Turns on transaction mode at the beginning. This improves performance. But if
the application crashes in the middle, all new changes will be lost
.RS
.RE

.TP 3
\fB-S\fR 
Turns Off synchronous mode.
With synchronous OFF, \fBmpdev_update\fR continues without syncing as soon
as it has handed data off to the operating system. In case \fBmpdev_update\fR 
crashes, the data will be safe, but the database might become corrupted if
the operating system crashes or the computer loses power before that data
has been written to the disk. On the other hand, commits can be orders of
magnitude faster with synchronous OFF.
.RS
.RE

.TP 3
\fB-P\fR 
Print SQL statements used during db creation
.RS
.RE

Examples
1. create stats.db in the current directory

.EX
$ mpdev_update -S -j -t -D 0 -d stats.db
Processed 42630 rows, Failures 0 rows, Updated 42636 rows

real    0m0.830s
user    0m0.405s
sys     0m0.096s
.EE

2. Update stats.db in the current directory and add 6 new songs
.EX
$ mpdev_update -S -j -t -D 0 -d stats.db
Processed 42636 rows, Failures 0 rows, Updated 6 rows

real    0m0.725s
user    0m0.353s
sys     0m0.067s
.EE

.SH REPORTING BUGS
.PP
If you find a bug, please report it at https://github.com/mbhangui/mpdev/issues

.SH COPYRIGHT
.PP
Copyright (c) 2020 Manvendra Bhangui <mpdev@indimail.org>
.PD 0
.P
.PD
Free
use of this software is granted under the terms of the GNU General
Public License (GPLv2).
.SH AUTHOR
Manvendra Bhangui <manvendra@indimail.org>

.SH SEE ALSO
mpd(1),
sqlite3(1),
mpdev_cleanup(1),
mpdev(1)


ChangeLog
*
Release 0.2 Start 02/03/2020
02/03/2020
01. Initial Release
02. added update_stats utility to maintain stats database
03. mpdev.c: set SONG_DURATION as seconds since epoch
04. player can now on its own add entries to stat database
14/07/2020
05. mpdev - handle empty playlist with no currentsong playing
06. added mpdev_update, mpdev_cleanup
15/07/2020
07. moved mpdev_update, mpdev_cleanup to libexec/mpdev
08. added mpdev_rename script
16/07/2020
09. mpdev_rename: added --query option
10. mpdev: use read instead of timeoutread
17/07/2020
11. mdpev: added START_TIME, END_TIME variables
12. player: refactored script to update ratings from rompr/sticker/stats
13. player: removed hardcored MySQL connection parameters
14. player: fixed variables getting clobbered by functions
19/07/2020
15. player: display sticker insert/updates after NewSong line
16. mpdev_update: added date_added field
17. mpdev_update: use st_mtime, instead of st_ctime for date_added field
18. mpdplaylist: use date_added field instead of last_modified
20/07/2020
19. mpdev: send status, currentsong command on player event to set
    ELAPSED_TIME
20. added rompr_hist play history utility for rompr
21. renamed rompr_hist to romprhist
21/07/2020
22. player: added get_rompr_playcount_ts() function to get playcount and
    timestamp from Playcounttable
23. player: option to run without sticker, stats enabled
24. player: create temp file for recording original start time
25. player: added error condition for getting rating from Ratingtable.
26. player: BUG. fixed rating tmp filename
27. player: display rating for RompR in logs in RompR scale
22/07/2020
28. added transfer_play
29. player: use AUTO_RATING env variable for automatic rating of unrated songs
30. create_service: fix for raspberrypi, dietpi
24/07/2020
31. mpdev_update: removed requirement of -m option for update mode
32. player: exit end-song if END_TIME is not set
33. player: fixed display playcount number in verbose output
28/07/2020
34. mpdev_cleanup.c: made -m optional
35. mpdev_update.c: added updation of Disc tag
35. player.in: added zztimestamp filed to Ratingtable
02/08/2020
36. set default values for play_count, rating, date_added fields.
37. use %mtime% tag from mpd datatabase for date_added filed instead of using stat(2) call
03/08/2020
38. mpdplaylist: added --host option to specify mpd host
39. mpdplaylist: added --shuffle option to shuffle playlist
05/08/2020
40. mpdplaylist: renamed --save option to --playlist
41. mpdev_rename: use SQL REPLACE instead of sed
06/08/2020
42. mpdplaylist: renamed --dummy option to --query
08/08/2020
42. mpdev.c, mpdev_cleanup.c: fixed usage/error strings
43. tcpopen.c: added missing return on connect() failure


mpdev_update.c
/*
 * $Log: mpdev_update.c,v $
 * Revision 1.4  2020-08-02 09:11:14+05:30  Cprogrammer
 * set default values for play_count, rating, date_added fields.
 * use %mtime% tag from mpd datatabase for date_added filed instead of using stat(2) call
 *
 * Revision 1.3  2020-07-28 12:42:03+05:30  Cprogrammer
 * added updation of Disc tag
 *
 * Revision 1.2  2020-07-24 09:41:15+05:30  Cprogrammer
 * removed requirement of -m option for update mode
 *
 * Revision 1.1  2020-07-19 18:15:46+05:30  Cprogrammer
 * Initial revision
 *
 *
 *
 * mpdev_update - utility for maintaining stats/sticker sqlite db
 *
 * Full documentation of mpd commands at https://www.musicpd.org/doc/html/protocol.html
 *
 * Optimization tips from
 * https://www.whoishostingthis.com/compare/sqlite/optimize/
 * https://www.wassen.net/sqlite-c.html
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_TIME_H
#include <time.h>
#endif
#ifdef HAVE_SQLITE3
#include <sqlite3.h>
#endif
#include <timeoutread.h>
#include <timeoutwrite.h>
#include <substdio.h>
#include <getln.h>
#include <stralloc.h>
#include <fmt.h>
#include <env.h>
#include <error.h>
#include <strerr.h>
#include <sgetopt.h>
#include <scan.h>
#include <str.h>
#include <env.h>
#include "tcpopen.h"

#ifndef	lint
static char     sccsid[] = "$Id: mpdev_update.c,v 1.4 2020-08-02 09:11:14+05:30 Cprogrammer Exp mbhangui $";
#endif

extern char    *strptime(const char *, const char *, struct tm *);
ssize_t         safewrite(int, char *, int);
ssize_t         saferead(int, char *, int);

substdio        mpdin, mpdout, ssout, sserr;
static stralloc uri = {0}, last_modified = {0}, album = {0}, artist = {0},
				date = {0}, genre = {0}, title = {0}, track = {0},
				disc = {0}, duration = {0}, line = {0};
char            strnum[FMT_ULONG];
int             timeout = 1200, verbose, db_type = -1, do_update = 0;
static sqlite3 *db;
char           *usage =
				"usage: mpdev_update [-i IP/Host | -s unix_socket] [-p port]\n"
				" -i IP     - IP address of MPD host. default 127.0.0.1\n"
				" -p port   - MPD listening port. default 6600\n"
				" -s socket - unix domain socket path\n"
				" -d        - sqlite3 dbfile path\n"
				" -j        - Enable Journal in Memory\n"
				" -S        - Enable Synch Mode\n"
				" -P        - Print sql statment\n"
				" -t        - Enable Transaction Mode\n"
				" -D 0|1    - insert stats | insert sticker (insert new records)\n"
				" -U        - update stats (insert new records + update old records)\n"
				" -v        - verbose output";

void
die_write(char *arg)
{
	if (db)
		sqlite3_close(db);
	substdio_puts(&sserr, "Requested action aborted: write error");
	if (arg) {
		substdio_put(&sserr, ": ", 2);
		substdio_puts(&sserr, arg);
	}
	substdio_put(&sserr, "\n", 1);
	substdio_flush(&sserr);
	_exit(111);
}

void
out(char *s)
{
	if (substdio_puts(&ssout, s) == -1)
		die_write(0);
}

void
flush()
{
	if (substdio_flush(&ssout) == -1)
		die_write("stdout");
}

void
err_out(char *s)
{
	if (substdio_puts(&sserr, s) == -1)
		die_write(0);
}

void
err_flush()
{
	if (substdio_flush(&sserr) == -1)
		die_write("sserr");
}

void
die_read(char *arg)
{
	if (db)
		sqlite3_close(db);
	if (arg) {
		substdio_put(&sserr, "Requested action aborted: ", 26);
		substdio_puts(&sserr, arg);
		substdio_put(&sserr, ": read error\n", 13);
	} else
		substdio_puts(&sserr, "Requested action aborted: read error\n");
	substdio_flush(&sserr);
	_exit(111);
}

void
die_alarm()
{
	if (db)
		sqlite3_close(db);
	substdio_puts(&sserr, "Requested action aborted: timeout\n");
	substdio_flush(&sserr);
	_exit(111);
}

void
die_nomem()
{
	if (db)
		sqlite3_close(db);
	substdio_puts(&sserr, "mpdev_update: out of memory\n");
	substdio_flush(&sserr);
	_exit(111);
}

ssize_t
saferead(int fd, char *buf, int len)
{
	int             r;

	if ((r = timeoutread(timeout, fd, buf, len)) == -1) {
		if (errno == error_timeout)
			die_alarm();
	}
	if (r < 0)
		die_read(0);
	return r;
}

ssize_t
safewrite(int fd, char *buf, int len)
{
	int             r;

	if ((r = timeoutwrite(timeout, fd, buf, len)) <= 0)
		die_write(0);
	return r;
}

void
print_missing(char *uri, char *what)
{
	err_out("missing information in uri: ");
	err_out(uri);
	err_out(": ");
	err_out(what);
	err_out("\n");
	err_flush();
}

void
print_song()
{
	if (verbose) {
		out("uri: [");
		out(uri.s);
	}
	if (last_modified.len) {
		if (verbose) {
			out("], ");
			out("Last-Modified: [");
			out(last_modified.s);
		}
	} else
		print_missing(uri.s, "Last-Modified");
	if (duration.len) {
		if (verbose) {
			out("], ");
			out("Duration: [");
			out(duration.s);
		}
	} else
		print_missing(uri.s, "Duration");
	if (title.len) {
		if (verbose) {
			out("], ");
			out("Title: [");
			out(title.s);
		}
	} else
		print_missing(uri.s, "Title");
	if (artist.len) {
		if (verbose) {
			out("], ");
			out("Artist: [");
			out(artist.s);
		}
	} else
		print_missing(uri.s, "Artist");
	if (album.len) {
		if (verbose) {
			out("], ");
			out("Album: [");
			out(album.s);
		}
	} /*else
		print_missing(uri.s, "Album"); */
	if (track.len) {
		if (verbose) {
			out("], ");
			out("Track: [");
			out(track.s);
		}
	} /*else
		print_missing(uri.s, "Track"); */
	if (disc.len) {
		if (verbose) {
			out("], ");
			out("Disc: [");
			out(disc.s);
		}
	} /*else
		print_missing(uri.s, "Disc"); */
	if (genre.len) {
		if (verbose) {
			out("], ");
			out("Genre: [");
			out(genre.s);
		}
	} /*else
		print_missing(uri.s, "Genre"); */
	if (date.len) {
		if (verbose) {
			out("], ");
			out("Date: [");
			out(date.s);
		}
	} /*else
		print_missing(uri.s, "Date"); */
	if (verbose) {
		out("]\n");
		flush();
	}
	return;
}

sqlite3_stmt   *
stats_database_init(char *database, int synch_mode, int journal_in_memory, int transaction_mode)
{
	char           *sql, *err_msg;
	sqlite3_stmt   *res;

	if (sqlite3_open(database, &db) != SQLITE_OK)
		strerr_die3x(111, database, ": ", (char *) sqlite3_errmsg(db));
	if (!synch_mode) {
		if (verbose) {
			out("PRAGMA synchronous  = OFF\n");
			flush();
		}
		if (sqlite3_exec(db, "PRAGMA synchronous = OFF", NULL, NULL, &err_msg) != SQLITE_OK) {
			sqlite3_close(db);
			strerr_die3x(111, database, ": PRAGMA synchronous = OFF: ", err_msg);
		}
		if (err_msg)
			sqlite3_free(err_msg);
	} else
	if (verbose) {
		out("PRAGMA synchronous  = NORMAL\n");
		flush();
	}
	if (journal_in_memory) {
		if (verbose) {
			out("PRAGMA journal_mode = MEMORY\n");
			flush();
		}
		if (sqlite3_exec(db, "PRAGMA journal_mode = MEMORY", NULL, NULL, &err_msg) != SQLITE_OK) {
			sqlite3_close(db);
			strerr_die3x(111, database, ": PRAGMA journal_mode = MEMORY: ", err_msg);
		}
		if (err_msg)
			sqlite3_free(err_msg);
	} else
	if (verbose) {
		out("PRAGMA journal_mode = DELETE\n");
		flush();
	}
	if (transaction_mode) {
		if (verbose) {
			out("BEGIN TRANSACTION\n");
			flush();
		}
		if (sqlite3_exec(db, "BEGIN TRANSACTION", NULL, NULL, &err_msg) != SQLITE_OK) {
			sqlite3_close(db);
			strerr_die3x(111, database, ": BEGIN TRANSACTION: ", err_msg);
		}
		if (err_msg)
			sqlite3_free(err_msg);
	}
	if (db_type == 0) {
		sql = "DROP INDEX IF EXISTS rating;\n"
			  "DROP INDEX IF EXISTS uri;\n"
			  "DROP INDEX IF EXISTS last_played;\n"
			  "DROP INDEX IF EXISTS date_added;\n"
			  "DROP INDEX IF EXISTS last_modified;";
	} else {
		sql = "DROP INDEX IF EXISTS rating;\n"
			  "DROP INDEX IF EXISTS uri;\n";
	}
	if (sqlite3_exec(db, sql, 0, 0, &err_msg) != SQLITE_OK) {
		sqlite3_close(db);
		strerr_die5x(111, database, ": ", sql, ": ", err_msg);
	}
	if (verbose) {
		if (sqlite3_changes(db) == 1) {
			out(sql);
			out("\n");
			flush();
		}
	}
	if (err_msg)
		sqlite3_free(err_msg);
	if (db_type == 0) {
		sql = "CREATE TABLE IF NOT EXISTS song(\n"
        	"id              INTEGER PRIMARY KEY,\n"
			"play_count      INTEGER DEFAULT 0,\n"
			"rating          INTEGER DEFAULT 0,\n"
        	"uri             TEXT UNIQUE NOT NULL,\n"
        	"duration        INTEGER,\n"
        	"last_modified   INTEGER,\n"
			"date_added      INTEGER DEFAULT (strftime('%s','now')),\n"
        	"artist          TEXT,\n"
        	"album           TEXT,\n"
        	"title           TEXT,\n"
        	"track           TEXT,\n"
        	"name            TEXT,\n"
        	"genre           TEXT,\n"
        	"date            TEXT,\n"
        	"composer        TEXT,\n"
        	"performer       TEXT,\n"
        	"disc            TEXT,\n"
        	"last_played     INTEGER,\n"
        	"karma           INTEGER NOT NULL CONSTRAINT karma_percent CHECK (karma >= 0 AND karma <= 100) DEFAULT 50\n"
		");";
	} else {
		sql = "CREATE TABLE IF NOT EXISTS sticker(\n"
        	"type  VARCHAR NOT NULL,\n"
        	"uri   VARCHAR NOT NULL,\n"
        	"name  VARCHAR NOT NULL,\n"
        	"value VARCHAR NOT NULL\n"
		");";
	}
	if (sqlite3_exec(db, sql, 0, 0, &err_msg) != SQLITE_OK) {
		sqlite3_close(db);
		strerr_die5x(111, database, ": ", sql, ": ", err_msg);
	}
	if (verbose) {
		if (sqlite3_changes(db) == 1) {
			out(sql);
			out("\n");
			flush();
		}
	}
	if (err_msg)
		sqlite3_free(err_msg);
	if (db_type == 0) {
		if (do_update == 0) /*- insert */
			sql = "INSERT or IGNORE into song (uri, artist, album, title, track, "
				"genre, date, date_added, last_modified, duration) "
			  	"values (@uri, @artist, @album, @title, @track, @genre, @date, "
				"@added, @modified, @duration)";
		else /*- update stats for any metadata updation */
			sql = "UPDATE song set artist=@artist, album=@album, title=@title, "
				"track=@track, genre=@genre, date=@date, last_modified=@modified, "
				"duration=@duration where uri=@uri";
	} else
		sql = "INSERT or IGNORE into sticker (type, uri, name, value)"
			  " values ('song', @uri, 'rating', 0)";
	if (sqlite3_prepare_v2(db, sql, -1, &res, 0) != SQLITE_OK) {
		sqlite3_close(db);
		strerr_die5x(111, database, ": sqlite3_prepare_v2: ", sql, ": ", (char *) sqlite3_errmsg(db));
	}
	return (res);
}

void
insert_update_data(sqlite3_stmt *res, unsigned long *processed, unsigned long *failure, char **ptr)
{
	struct tm       tm = {0};
	int             i;
	time_t          mod_time;

	if (db_type == 0) {
		if (!strptime(last_modified.s, "%Y-%m-%dT%H:%M:%SZ", &tm)) {
			strerr_warn4("uri: ", uri.s, ": invalid timestamp: ", last_modified.s, 0);
			return;
		} else
		if (!(mod_time = mktime(&tm))) {
			strerr_warn4("uri: ", uri.s, ": invalid timestamp: ", last_modified.s, 0);
			return;
		}
		if (do_update == 0) { /*- insert */
			sqlite3_bind_text(res,  1, uri.len ? uri.s : "", -1, SQLITE_STATIC);
			sqlite3_bind_text(res,  2, artist.len ? artist.s : "", -1, SQLITE_STATIC);
			sqlite3_bind_text(res,  3, album.len ? album.s : "", -1, SQLITE_STATIC);
			sqlite3_bind_text(res,  4, title.len ? title.s : "", -1, SQLITE_STATIC);
			sqlite3_bind_text(res,  5, track.len ? track.s : "", -1, SQLITE_STATIC);
			sqlite3_bind_text(res,  6, genre.len ? genre.s : "", -1, SQLITE_STATIC);
			sqlite3_bind_text(res,  7, date.len ? date.s : "", -1, SQLITE_STATIC);
			sqlite3_bind_int(res,   8, mod_time);
			sqlite3_bind_int(res,   9, mod_time);
			sqlite3_bind_text(res, 10, duration.s, -1, SQLITE_STATIC);
		} else { /*- update stats for any metadata updation */
			i = 1;
			sqlite3_bind_text(res, i++, artist.len ? artist.s : "", -1, SQLITE_STATIC);
			sqlite3_bind_text(res, i++, album.len ? album.s : "", -1, SQLITE_STATIC);
			sqlite3_bind_text(res, i++, title.len ? title.s : "", -1, SQLITE_STATIC);
			sqlite3_bind_text(res, i++, track.len ? track.s : "", -1, SQLITE_STATIC);
			sqlite3_bind_text(res, i++, genre.len ? genre.s : "", -1, SQLITE_STATIC);
			sqlite3_bind_text(res, i++, date.len ? date.s : "", -1, SQLITE_STATIC);
			sqlite3_bind_int(res,  i++, mod_time);
			sqlite3_bind_text(res, i++, duration.s, -1, SQLITE_STATIC);
			sqlite3_bind_text(res, i++, uri.len ? uri.s : "", -1, SQLITE_STATIC);
			/*-
			sql = "UPDATE song set artist=@artist, album=@album, title=@title, track=@track, "
				"genre=@genre, date=@date, date_added=@added, last_modified=@modified, "
				"duration=@duration where uri=@uri";
			*/
		}
	} else
		sqlite3_bind_text(res, 1, uri.len ? uri.s : "", -1, SQLITE_STATIC);
	if (ptr)
		*ptr = sqlite3_expanded_sql(res);
	if (sqlite3_step(res) != SQLITE_DONE) {
		*failure += 1;
	} else
		*processed += 1;
	sqlite3_clear_bindings(res);
	sqlite3_reset(res);
}

int
stats_database_end(char *database, sqlite3_stmt *res, int transaction_mode)
{
	char           *sql, *err_msg;
	int             i;

	if (db_type == 0) {
		sql = "CREATE INDEX IF NOT EXISTS rating        on song(rating);\n"
			  "CREATE INDEX IF NOT EXISTS uri           on song(uri);\n"
			  "CREATE INDEX IF NOT EXISTS last_played   on song(last_played);\n"
			  "CREATE INDEX IF NOT EXISTS date_added    on song(date_added);\n"
			  "CREATE INDEX IF NOT EXISTS last_modified on song(last_modified);";
	} else {
		sql = "CREATE INDEX IF NOT EXISTS rating        on sticker(value);\n"
			  "CREATE INDEX IF NOT EXISTS uri           on sticker(uri);\n";
	}
	if (sqlite3_exec(db, sql, 0, 0, &err_msg) != SQLITE_OK) {
		sqlite3_close(db);
		strerr_die5x(111, database, ": ", sql, ": ", err_msg);
	}
	if (verbose) {
		if (sqlite3_changes(db) == 1) {
			out(sql);
			out("\n");
			flush();
		}
	}
	if (err_msg)
		sqlite3_free(err_msg);
	if (transaction_mode) {
		if (verbose) {
			out("END TRANSACTION\n");
			flush();
		}
		if (sqlite3_exec(db, "END TRANSACTION", 0, 0, &err_msg) != SQLITE_OK) {
			sqlite3_close(db);
			strerr_die3x(111, database, ": end transaction: ", err_msg);
		}
		if (err_msg)
			sqlite3_free(err_msg);
	}
	i = sqlite3_total_changes(db);
	if (sqlite3_finalize(res) != SQLITE_OK)
		strerr_warn2("uri: ", "sqlite3_finalize: ", (char *) sqlite3_errmsg(db));
	sqlite3_close(db);
	return (i);
}

int
main(int argc, char **argv)
{
	int             opt, match, sock, port_num = 6600, is_directory = 0;
	char            synch_mode = 1, journal_in_memory = 0, transaction_mode = 0, print_sql = 0;
	unsigned long   processed, failure, row_count;
	time_t          t;
	struct tm       tm = {0};
	char            port[FMT_ULONG], mpdinbuf[1024], mpdoutbuf[512], ssoutbuf[512], sserrbuf[512];
	char           *mpd_socket, *mpd_host, *ptr, *database;
	sqlite3_stmt   *res;

	substdio_fdbuf(&ssout, write, 1, ssoutbuf, sizeof(sserrbuf));
	substdio_fdbuf(&sserr, write, 2, sserrbuf, sizeof(sserrbuf));
	if (!(mpd_host = env_get("MPD_HOST")))
		mpd_host = "127.0.0.1";
	mpd_socket = env_get("MPD_SOCKET");
	if (!(ptr = env_get("MPD_PORT")))
		port_num = 6600;
	else
		scan_uint(ptr, (unsigned int *) &port_num);
	database = (char *) 0;
	while ((opt = getopt(argc, argv, "vjSUPth:p:s:d:D:")) != opteof) {
		switch (opt) {
		case 'd':
			database = optarg;
			break;
		case 'h':
			mpd_host = optarg;
			break;
		case 's':
			mpd_socket = optarg;
			break;
		case 'p':
			scan_uint(optarg, (unsigned int *) &port_num);
			break;
		case 'j':
			journal_in_memory = 1;
			break;
		case 'S':
			synch_mode = 0;
			break;
		case 't':
			transaction_mode = 1;
			break;
		case 'P':
			print_sql = 1;
			break;
		case 'D':
			scan_uint(optarg, (unsigned int *) &db_type);
			break;
		case 'U':
			do_update = 1;
			break;
		case 'v':
			verbose++;
			break;
		default:
			strerr_die1x(100, usage);
		}
	}
	t = time(0);
	localtime_r(&t, &tm);
	if (!env_put2("TZ", (char *) tm.tm_zone))
		die_nomem();
	if (mpd_socket && port_num != 6600)
		strerr_die1x(100, "you can't specify both socket & port");
	if (mpd_socket && str_diff(mpd_host, "127.0.0.1"))
		strerr_die1x(100, "you can't specify both socket & IP");
	if (!database) {
		strerr_warn1("mpdev_update: database (-d option) not specified", 0);
		strerr_die1x(100, usage);
	}
	if (db_type == -1) {
		strerr_warn1("mpdev_update: db type (-D option) not specified", 0);
		strerr_die1x(100, usage);
	}
	port[fmt_ulong(port, port_num)] = 0;
	if ((sock = tcpopen(mpd_socket ? mpd_socket : mpd_host, 0, port_num)) == -1) {
		if (mpd_socket)
			strerr_die4sys(111, "mpdev_update: tcpopen: ", "socket [", mpd_socket, "]");
		else
			strerr_die6sys(111, "mpdev_update: tcpopen: ", "host [", mpd_host, "] port [", port, "]");
	}
	substdio_fdbuf(&mpdin, saferead, sock, mpdinbuf, sizeof mpdinbuf);
	substdio_fdbuf(&mpdout, safewrite, sock, mpdoutbuf, sizeof mpdoutbuf);
	if (substdio_put(&mpdout, "listallinfo\n", 12) || substdio_flush(&mpdout) == -1)
		die_write("unable to write to mpd");
	res = stats_database_init(database, synch_mode, journal_in_memory, transaction_mode);
	uri.len = last_modified.len = duration.len = artist.len = album.len = title.len = track.len = genre.len = date.len = 0;
	for (processed = failure = 0;;) {
		if (getln(&mpdin, &line, &match, '\n') == -1)
			die_read("getln");
		if (!match && line.len == 0)
			break;
		if (!str_diffn(line.s, "OK\n", 3))
			break;
		line.s[--line.len] = 0; /*- remove newline */
		if (!str_diffn(line.s, "directory: ", 11))
			is_directory = 1;
		if (!str_diffn(line.s, "file: ", 6)) {
			is_directory = 0;
			if (uri.len) {
				insert_update_data(res, &processed, &failure, print_sql ? &ptr : 0);
				if (sqlite3_changes(db) == 1) {
					print_song(uri.s, last_modified.s, duration.s, artist.s, album.s, title.s, track.s, genre.s, date.s);
					if (print_sql && ptr) {
						out(ptr);
						sqlite3_free(ptr);
						out("\n");
						flush();
					}
				}
				uri.len = last_modified.len = duration.len = artist.len = album.len = title.len = track.len = genre.len = date.len = 0;
			}
			if (!stralloc_copyb(&uri, line.s + 6, line.len - 5))
				die_nomem();
			uri.len--;
		}
		if (is_directory)
			continue;
		if (!str_diffn(line.s, "Last-Modified: ", 15)) {
			if (!stralloc_copyb(&last_modified, line.s + 15, line.len - 14))
				die_nomem();
			last_modified.len--;
		} else
		if (!str_diffn(line.s, "Time: ", 6)) {
			if (!stralloc_copyb(&duration, line.s + 6, line.len - 5))
				die_nomem();
			duration.len--;
		} else
		if (!str_diffn(line.s, "Title: ", 7)) {
			if (!stralloc_copyb(&title, line.s + 7, line.len - 6))
				die_nomem();
			title.len--;
		} else
		if (!str_diffn(line.s, "Artist: ", 8)) {
			if (!stralloc_copyb(&artist, line.s + 8, line.len - 7))
				die_nomem();
			artist.len--;
		} else
		if (!str_diffn(line.s, "Album: ", 7)) {
			if (!stralloc_copyb(&album, line.s + 7, line.len - 6))
				die_nomem();
			album.len--;
		} else
		if (!str_diffn(line.s, "Track: ", 7)) {
			if (!stralloc_copyb(&track, line.s + 7, line.len - 6))
				die_nomem();
			track.len--;
		} else
		if (!str_diffn(line.s, "Disc: ", 6)) {
			if (!stralloc_copyb(&disc, line.s + 6, line.len - 5))
				die_nomem();
			disc.len--;
		} else
		if (!str_diffn(line.s, "Genre: ", 7)) {
			if (!stralloc_copyb(&genre, line.s + 7, line.len - 6))
				die_nomem();
			genre.len--;
		} else
		if (!str_diffn(line.s, "Date: ", 6)) {
			if (!stralloc_copyb(&date, line.s + 6, line.len - 5))
				die_nomem();
			date.len--;
		}
	}
	row_count = stats_database_end(database, res, transaction_mode);
	out("Processed ");
	strnum[fmt_ulong(strnum, processed)] = 0;
	out(strnum);
	strnum[fmt_ulong(strnum, failure)] = 0;
	out(" rows, Failures ");
	out(strnum);
	strnum[fmt_ulong(strnum, row_count)] = 0;
	out(" rows, Updated ");
	out(strnum);
	out(" rows\n");
	flush();
	close(sock);
	return 0;
}
